src/Container.js

```diff
@@ -0,0 +1,334 @@
+    this.reset(); // force close the app
+  }
+
+  /**
+   * Open a application or path
+   * @param {string} userPath The full path to the application to load
+   * @param {Object} [userOptions] The open options
+   * @param {Boolean} [userOptions.singlePlay=false] If we should play in single play mode
+   * @param {Object | null} [userOptions.playOptions=null] The optional play options
+   * @memberof Container
+   */
+  _internalOpen(userPath, { singlePlay = false, playOptions = null } = {}) {
+    const options = { singlePlay, playOptions };
+    this.reset();
+
+    const err = Features.basic(); // TODO Features.basic()
+    if (err) {
+      return this.client.trigger('unsupported');
+    }
+
+    this.loading = true;
+
+    this.initClient();
+
+    this.plugins.forEach(plugin => plugin.open(this));
+
+    let path = userPath;
+    if (null !== options.playOptions) {
+      const playOptionsQueryString =
+        'playOptions=' +
+        encodeURIComponent(JSON.stringify(options.playOptions));
+
+      path =
+        -1 === userPath.indexOf('?')
+          ? `${userPath}?${playOptionsQueryString}`
+          : `${userPath}&${playOptionsQueryString}`;
+    }
+
+    this.main.classList.add('loading');
+    this.main.setAttribute('src', path);
+
+    this.client.respond('singlePlay', options.singlePlay);
+    this.client.respond('playOptions', options.playOptions);
+    this.client.trigger('open');
+  }
+
+  /**
+   *
+   *
+   * @param {*} path
+   * @param {*} [options={}]
+   * @param {*} [playOptions={}]
+   * @memberof Container
+   */
+  openPath(path, options = {}, playOptions = {}) {
+    // This should be deprecated, support for old function signature
+    if ('boolean' === typeof options) {
+      options = {
+        singlePlay: false,
+        playOptions: playOptions
+      };
+    }
+    this._internalOpen(path, options);
+  }
+
+  /**
+   * Destroy and don't use after this
+   * @memberof Container
+   */
+  destroy() {
+    this.reset();
+    // Destroy in the reverse priority order
+    this.plugins
+      .slice()
+      .reverse()
+      .forEach(plugin => plugin.teardown(this));
+
+    this.main = null;
+    this.options = null;
+    this.dom = null;
+  }
+
+  /**
+   * Tell the application to start closing
+   * @memberof Container
+   */
+  close() {
+    if (this.loading || this.loaded) {
+      this.plugins.forEach(plugin => plugin.close(this));
+      this.client.trigger('close');
+      // Start the close
+      this.client.send('close');
+    } else {
+      this.reset();
+    }
+  }
+
+  /**
+   * The current version of SpringRollContainer
+   * @readonly
+   * @static
+   * @memberof Container
+   */
+  static get version() {
+    return 'CONTAINER_VERSION';
+  }
+
+  /**
+   *
+   *
+   * @static
+   * @param {object|function} plugin your plugin. This will be merged with a base plugin to make sure your plugin has certain functions
+   * @memberof Container
+   */
+  static uses(plugin = BasePlugin) {
```

Comment:
I mentioned this earlier, but we'll want to revisit this. In this case, I'd rather pass the _instance_ of the plugin instead rather that the class. My reasoning for this is that the constructor to the plugin might be parameterized so developers would want to configure it ahead of time before calling use:

```javascript
const pluginInstance = new FancyPlugin({
  apiKey: 'abc123',
  debug: true
});

Container.uses(pluginInstance);
const container = new Container('#iframe');

container.openPath('asdf');
```

Reply:
Makes sense. I actually was playing around with the idea where you could pass a class, a instance of a class, function, or just object, but it got pretty convoluted and there was some scope issues. Might be worth to take another crack at that though.